clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```
@@ -848,6 +848,8 @@ public void handleResponse(AbstractResponse response) {
                         coordinatorReloaded = true;
                         lookupCoordinator(FindCoordinatorRequest.CoordinatorType.GROUP, builder.consumerGroupId());
                     }
+                } else if (error == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
```

Mentioned offline, I think the only thing I'm unsure of is whether it is useful to refresh metadata in this case. @guozhangwang What do you think?

---

Never mind. The issue here is that the broker potentially has stale metadata, not us, so refreshing wouldn't help.

---

Could we check that the `response.requestHeader.apiKey` is either `addPartitionsToTxn` or `txnCommitOffset`? For example `addOffsetsToTxn` should not return this error code and hence if yes we should fail fast as a bug.

---

@guozhangwang not sure I follow: Only the handlers for these two request types will treat an `UNKNOWN_TOPIC_OR_PARTITION` response as a retriable error. All the other handlers will mark the transaction in error state if this response occurs, and then the only option open to the application is to abort the transaction.

Do you mean to do additional validation that mapping of callback to request is indeed correct?

---

Nvm, I thought it is a general response handler but it is only for `txnCommitOffset`.